dialogue-apache-hc5-client/src/main/java/com/palantir/dialogue/hc5/ApacheHttpClientBlockingChannel.java

```diff
@@ -167,6 +154,42 @@ public Response execute(Endpoint endpoint, Request request) throws IOException {
         }
     }
 
+    @VisibleForTesting
+    static ClassicHttpRequest createRequest(BaseUrl baseUrl, Endpoint endpoint, Request request) {
+        // Create base request given the URL
+        URL target = baseUrl.render(endpoint, request);
+        ClassicRequestBuilder builder = ClassicRequestBuilder.create(
+                        endpoint.httpMethod().name())
+                .setScheme(target.getProtocol())
+                .setAuthority(parseAuthority(target))
+                .setPath(target.getFile());
+
+        // Fill headers
+        request.headerParams().forEach(builder::addHeader);
+
+        if (request.body().isPresent()) {
+            Preconditions.checkArgument(
+                    endpoint.httpMethod() != HttpMethod.GET, "GET endpoints must not have a request body");
+            Preconditions.checkArgument(
+                    endpoint.httpMethod() != HttpMethod.HEAD, "HEAD endpoints must not have a request body");
+            Preconditions.checkArgument(
+                    endpoint.httpMethod() != HttpMethod.OPTIONS, "OPTIONS endpoints must not have a request body");
+            setBody(builder, request.body().get());
+        } else if (requiresEmptyBody(endpoint)) {
+            builder.setEntity(EmptyHttpEntity.INSTANCE);
+        }
+        return builder.build();
+    }
+
+    @VisibleForTesting
+    static URIAuthority parseAuthority(URL url) {
+        try {
+            return URIAuthority.create(url.getAuthority());
+        } catch (URISyntaxException e) {
+            throw new SafeIllegalArgumentException("Invalid URI authority", e, UnsafeArg.of("url", url));
+        }
```

Comment:
I wonder if it's worth rolling the dice with something along these lines to avoid string tokenizing/manipulation overhead? We have most of the data parsed out via URL already, I believe it's just ipv6 brackets that URL doesn't handle well.

```java
String host = url.getHost();
if (host != null && host.startsWith("[") && host.endsWith("]")) {
    host = host.substring(1, host.length() - 1);
}
return new URIAuthority(url.getUserInfo(), host, url.getPort());
```

Reply:
Alternatively, it may be safer if we check `host != null && host.startsWith("[")` and in that case, we use the parsing `URIAuthority.create` path, otherwise (in the common case) use `new URIAuthority(url.getUserInfo(), host, url.getPort());`